### ISSUE TITLE
Missing schema for newer versions of Redis API

### ISSUE THREAD

**Issue by matthchr** [Azure/azure-resource-manager-schemas#1237]
The Microsoft.Cache API seems to be in the public Swagger repo but I don't see corresponding entries here in the 2015-01-01 rollup version.

There is a [2018-03-01](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/redis/resource-manager/Microsoft.Cache/stable) and a [2019-07-01-preview](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/redis/resource-manager/Microsoft.Cache/preview/2019-07-01) both missing.

**Comment by TimLovellSmith** [comment]
Looks like we need to get Microsoft.Cache onboarded for autogeneration.

**Comment by TimLovellSmith** [comment]
Or... maybe not. Microsoft.Cache is already in the list in https://github.com/Azure/azure-resource-manager-schemas/blob/master/generator/autogenlist.ts
So why would newer versions be missing?

**Comment by matthchr** [comment]
It may be because the 2015-01-01 spec is already so large that this whole API version was decided to be cut from it?

**Comment by TimLovellSmith** [comment]
@mattchr I don't really grasp the details of this. How would a spec or api-version be too large? Who would cut it? How?

**Comment by matthchr** [comment]
The schemas have to be parsed in order to be used, and as they grow larger that takes more time. @anthony-c-martin understands far better than I do but I think the result is that for the schemas that are already very large (2015-01-01) they are trying to avoid growing them even larger and they are doing that by omitting some API versions maybe?

**Comment by TimLovellSmith** [comment]
@matthchr I think I see. Well, given we have come along a couple years/versions and even have 2020 if not 2021 api versions now, I think its _probably_ about time for a new one by now... should we simultaneously retire any one ones?

**Comment by TimLovellSmith** [comment]
So while trying to figure out what this is really about, I came out quite confused about how this is all supposed to be structured. Are the deploymentTemplate schemas supposed to link to every other resource type schema? Which version? What if the version numbers are not aligned?

**Comment by TimLovellSmith** [comment]
(So would #1520 address this?)

**Comment by anthony-c-martin** [comment]
> So while trying to figure out what this is really about, I came out quite confused about how this is all supposed to be structured. Are the deploymentTemplate schemas supposed to link to every other resource type schema? Which version? What if the version numbers are not aligned?

Basically, it turns out that when using JSON schema 4 to validate or provide code completion functionality naively for resources, an exhaustive pattern match across 200+ files with multiple $refs each is **EXTREMELY** slow (talking 30+ seconds for some implementations I've seen). So the decision was made to avoid adding any new resource types to the existing root schemas (see note [here](https://github.com/Azure/azure-resource-manager-schemas#submitting-a-pr)), and to publish a new root schema (2019-04-01) which contains a complete list.

The ARM VSCode extension no longer uses the built-in schema processor, and instead uses custom logic to index resources by type & api-version for much more performant type lookup. The idea at the time was that choosing to use 2019-04-01 was a conscious decision that users of VSCode could choose to make, but we wouldn't 'break' any users of standard JSON schema parsers by suddenly tripling the number of resources with autogeneration.

2019-04-01 contains both a list of hand-authored schemas in-line, and a link to the set of automatically generated schemas here: https://github.com/Azure/azure-resource-manager-schemas/blob/master/schemas/common/autogeneratedResources.json.

**tl;dr**:
* If you are hand-authoring schemas, you need to keep [2019-04-01](https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json) up-to-date.
* If you are auto-generating schemas, you just need to add your config to the [autogenlist](https://github.com/Azure/azure-resource-manager-schemas/blob/master/generator/autogenlist.ts), and let the generator handle the rest.

**Comment by anthony-c-martin** [comment]
Looks like the reported missing schema versions are now present:
https://github.com/Azure/azure-resource-manager-schemas/blob/master/schemas/2018-03-01/Microsoft.Cache.json
https://github.com/Azure/azure-resource-manager-schemas/blob/master/schemas/2019-07-01/Microsoft.Cache.json

@mattchr are you still seeing problems here?

**Comment by matthchr** [comment]
@anthony-c-martin - No as far as I know things are fine here. I think that this can probably be closed.

**Comment by anthony-c-martin** [comment]
> @anthony-c-martin - No as far as I know things are fine here. I think that this can probably be closed.

Thanks for confirming!
